k9mail/src/test/java/com/fsck/k9/crypto/MessageDecryptVerifierTest.java

```diff
@@ -86,8 +91,206 @@ public void findEncryptedPartsShouldReturnMultipleEncryptedParts() throws Except
+                                bodypart("application/pdf")
+                        ),
+                        bodypart("application/pgp-signature")
+                )
+        );
+
+        List<Part> signedParts = MessageDecryptVerifier.findSignedParts(message);
+
+        assertEquals(1, signedParts.size());
+        assertSame(message, signedParts.get(0));
+    }
+
+    @Test
+    public void findEncrypted__withMultipartMixedSubSigned__shouldReturnSigned() throws Exception {
+        Message message = messageFromBody(
+                multipart("mixed",
+                    multipart("signed",
+                            bodypart("text/plain"),
+                            bodypart("application/pgp-signature")
+                    )
+                )
+        );
+
+        List<Part> signedParts = MessageDecryptVerifier.findSignedParts(message);
+
+        assertEquals(1, signedParts.size());
+        assertSame(createdMultiparts.get(0), signedParts.get(0));
+    }
+
+    @Test
+    public void findEncrypted__withMultipartMixedSubSignedAndText__shouldReturnSigned() throws Exception {
+        Message message = messageFromBody(
+                multipart("mixed",
+                        multipart("signed",
+                                bodypart("text/plain"),
+                                bodypart("application/pgp-signature")
+                        ),
+                        bodypart("text/plain")
+                )
+        );
+
+        List<Part> signedParts = MessageDecryptVerifier.findSignedParts(message);
+
+        assertEquals(1, signedParts.size());
+        assertSame(createdMultiparts.get(0), signedParts.get(0));
+    }
+
+    @Test
+    public void findEncrypted__withMultipartMixedSubTextAndSigned__shouldReturnSigned() throws Exception {
+        Message message = messageFromBody(
+                multipart("mixed",
+                        bodypart("text/plain"),
+                        multipart("signed",
+                                bodypart("text/plain"),
+                                bodypart("application/pgp-signature")
+                        )
+                )
+        );
+
+        List<Part> signedParts = MessageDecryptVerifier.findSignedParts(message);
+
+        assertEquals(1, signedParts.size());
+        assertSame(createdMultiparts.get(0), signedParts.get(0));
+    }
+
+    MimeMessage messageFromBody(BodyPart bodyPart) throws MessagingException {
+        MimeMessage message = new MimeMessage();
+        MimeMessageHelper.setBody(message, bodyPart.getBody());
+        return message;
+    }
+
+    MimeBodyPart multipart(String type, BodyPart... subParts) throws MessagingException {
+        MimeMultipart multiPart = new MimeMultipart();
+        multiPart.setSubType(type);
+        for (BodyPart subPart : subParts) {
+            multiPart.addBodyPart(subPart);
+        }
+        MimeBodyPart resultPart = new MimeBodyPart(multiPart);
+        createdMultiparts.add(resultPart);
```

Comment:
Instead of using `createdMultiparts` I think tests would be easier to read if `message` were used to retrieve a reference to the desired child part.
Maybe with a helper method that makes addressing parts easier, e.g.

```java
Part signaturePart = getPart(message, "1.1");
```

to get the `application/pgp-signature` from

```java
Message message = messageFromBody(
        multipart("mixed",
                bodypart("text/plain"),
                multipart("signed",
                        bodypart("text/plain"),
                        bodypart("application/pgp-signature")
                )
        )
);
```

(Not that we need a reference to the signed part. Just makes for a better example.)